### PR TITLE
Minor: improve performance of `ScalarValue::Binary*` debug

### PR DESCRIPTION
**Pull request by alamb**
## Rationale for this change
This is similar to https://github.com/apache/datafusion/pull/12322

(Another) Follow on to https://github.com/apache/datafusion/pull/12192 from @lewiszlw

## What changes are included in this PR?

1. Add tests for binary debug
2. Fix a bug that the tests found
3. improve performance with fewer allocations

## Are these changes tested?

Yes

### CODE REVIEW

File: datafusion/common/src/scalar/mod.rs [hunk 1]
```diff
@@ -3707,7 +3711,7 @@ impl fmt::Debug for ScalarValue {
                 write!(f, "FixedSizeBinary({size}, {self})")
             }
             ScalarValue::FixedSizeBinary(size, Some(b)) => {
-                write!(f, "Binary({size}, \"")?;
+                write!(f, "FixedSizeBinary({size}, \"")?;
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:3714]: this is a bug (likely copy/paste error) found with the tests

File: datafusion/common/src/scalar/mod.rs [hunk 2]
```diff
@@ -3646,18 +3646,22 @@ fn fmt_list(arr: ArrayRef, f: &mut fmt::Formatter) -> fmt::Result {
     write!(f, "{value_formatter}")
 }
 
+/// writes a byte array to formatter. `[1, 2, 3]` ==> `"1,2,3"`
+fn fmt_binary(data: &[u8], f: &mut fmt::Formatter) -> fmt::Result {
+    let mut first = true;
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:3651]: the difference here is that `join` allocates a new `String`

**Review by lewiszlw — APPROVED** [review]
👍

File: datafusion/common/src/scalar/mod.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/common/src/scalar/mod.rs:3651]: ... and `format!` as well :)

File: datafusion/common/src/scalar/mod.rs [hunk 3]
```diff
@@ -3646,18 +3646,22 @@ fn fmt_list(arr: ArrayRef, f: &mut fmt::Formatter) -> fmt::Result {
...
+    for b in data {
```
Comment (Dandandan) [inline datafusion/common/src/scalar/mod.rs:3652]: I think this is slightly more idiomatic written as follows:
```
fn fmt_binary(data: &[u8], f: &mut fmt::Formatter) -> fmt::Result {
    let mut data_iter = data.iter();
    if let Some(b) = data_iter.next() {
        write!(f, "{}", b)?;
    }
    for b in data_iter {
        write!(f, ",{b}")?;
    }
    Ok(())
}
```

**Review by Dandandan — APPROVED** [review]

**Comment by Dandandan** [comment]
Added one small suggestion

File: datafusion/common/src/scalar/mod.rs (hunk 3 above)
Comment (Dandandan) [reply datafusion/common/src/scalar/mod.rs:3652]: Or even better
```
fn fmt_binary(data: &[u8], f: &mut fmt::Formatter) -> fmt::Result {
    if let Some(b) = data.first() {
        write!(f, "{}", b)?;
    }
    for b in data.iter().skip(1) {
        write!(f, ",{b}")?;
    }
    Ok(())
}
```

Comment (alamb) [reply datafusion/common/src/scalar/mod.rs]: I came up with something slightly different that I still think is more idiomatic in  https://github.com/apache/datafusion/pull/12323/commits/99a57e77b0ed6641982af65d8a33ef9b9a73296f.

```rust
/// writes a byte array to formatter. `[1, 2, 3]` ==> `"1,2,3"`
fn fmt_binary(data: &[u8], f: &mut fmt::Formatter) -> fmt::Result {
    let mut iter = data.iter();
    if let Some(b) = iter.next() {
        write!(f, "{b}")?;
    }
    for b in iter {
        write!(f, ",{b}")?;
    }
    Ok(())
}
```

**Pull request merged**